File: langs/StdLibs/one.php
<?php

class OneRegex {
    static function matchFromIndex($pattern, $input, $offset) {
        preg_match('/\G'.str_replace('/', '\/', $pattern).'/', $input, $matches, 0, $offset);
        return $matches;
    }
}

class OneReflect {
    static $classes = array();

    static function getClass($obj) { 
        return OneReflect::getClassByName(get_class($obj));
    }

    static function getClassByName($name) {
        $key = OneReflect::nameKey($name);
        return array_key_exists($key, OneReflect::$classes) ? OneReflect::$classes[$key] : NULL;
    }

    static function setupClass($cls) {
        OneReflect::$classes[OneReflect::nameKey($cls->name)] = $cls;
    }

    static function nameKey($name) {
        return str_replace("_", "", strtolower($name));
    }
}

class OneClass {
    function __construct($name, $fields, $methods) {
        $this->name = $name;

        $this->fields = array();
        foreach ($fields as $field) {
            $field->cls = $this;
            $this->fields[OneReflect::nameKey($field->name)] = $field;
        }

        $this->methods = array();
        foreach ($methods as $method) {
            $method->cls = $this;
            $this->methods[OneReflect::nameKey($method->name)] = $method;
        }
    }

    function getField($name) { return $this->fields[OneReflect::nameKey($name)]; }
    function getMethod($name) { return $this->methods[OneReflect::nameKey($name)]; }
    function getFields() { return array_values($this->fields); }
    function getMethods() { return array_values($this->methods); }
}

class OneField {
    function __construct($name, $isStatic, $type) {
        $this->name = $name;
        $this->isStatic = $isStatic;
        $this->type = $type;
    }

    function getValue($obj) {
        if ($this->isStatic) {
            $field = $this->name;
            $obj = $this->cls->name;
            return $obj::$$field;
        } else {
            return $obj->{$this->name};
        }
    }

    function setValue($obj, $value) {
        if ($this->isStatic) {
            $field = $this->name;
            $obj = $this->cls->name;
            $obj::$$field = $value;
        } else {
            $obj->{$this->name} = $value;
        }
    }
}

class OneMethod {
    function __construct($name, $isStatic, $returnType, $args) {
        $this->name = $name;
        $this->isStatic = $isStatic;
        $this->returnType = $returnType;
        $this->args = $args;
    }

    function call($obj, $args) {
        $gotArgCount = count($args);
        $expectedArgCount = count($this->args);
        if ($gotArgCount !== $expectedArgCount)
            throw new Exception("Expected {$expectedArgCount} arguments, but got {$gotArgCount} in {$this->cls->name}::{$this->name} call!");

        $realObj = $this->isStatic ? $this->cls->name : $obj;
        return call_user_func_array(array($realObj, $this->name), $args);
    }
}

class OneMethodArgument {
    function __construct($name, $type) {
        $this->name = $name;
        $this->type = $type;
    }
}

File: langs/StdLibs/one.pl
package One;

sub str_replace
{
    my ($string, $find, $replace) = @_;

    my $pos = index($string, $find);

    while($pos > -1) {
        substr($string, $pos, length($find), $replace);
        $pos = index($string, $find, $pos + length($replace));
    }
    
    return $string;
}

package OneRegex;

sub match_from_index
{
    my ($pattern, $input, $offset) = @_;
    pos($input) = $offset;
    my @matches = ($input =~ /\G$pattern/g);
    my $result = scalar(@matches) ? \@matches : undef;
    return $result;
}

package OneClass;

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    my ( $name, $fields, $methods ) = @_;

    $self->{name} = $name;

    $self->{fields} = {};
    foreach my $field (@{$fields}) {
        $field->{cls} = $self;
        my $key = OneReflect::name_key($field->{name});
        $self->{fields}{$key} = $field;
    }

    $self->{methods} = {};
    foreach my $method (@{$methods}) {
        $method->{cls} = $self;
        my $key = OneReflect::name_key($method->{name});
        $self->{methods}{$key} = $method;
    }
    
    return $self;
}

sub get_field
{
    my ( $self, $name ) = @_;
    my $key = OneReflect::name_key($name);
    return $self->{fields}{$key};
}

sub get_method
{
    my ( $self, $name ) = @_;
    my $key = OneReflect::name_key($name);
    return $self->{methods}{$key};
}

sub get_fields
{
    my ( $self ) = @_;
    return values(@{$self->fields});
}
sub get_methods
{
    my ( $self ) = @_;
    return values(@{$self->methods});
}

package OneField;

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    my ( $name, $isStatic, $type ) = @_;
    $self->{name} = $name;
    $self->{isStatic} = $isStatic;
    $self->{type} = $type;
    return $self;
}

sub get_value
{
    my ( $self, $obj ) = @_;
    my $fieldName = $self->{name};
    if ($self->{isStatic}) {
        $className = $self->{cls}->{name};
        return ${"$className::$fieldName"};
    } else {
        return $obj->{$fieldName};
    }
}

sub set_value
{
    my ( $self, $obj, $value ) = @_;
    my $fieldName = $self->{name};
    if ($self->{isStatic}) {
        $className = $self->{cls}->{name};
        ${"$className::$fieldName"} = $value;
    } else {
        $obj->{$fieldName} = $value;
    }
}

package OneMethod;

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    my ( $name, $isStatic, $returnType, $args ) = @_;
    $self->{name} = $name;
    $self->{isStatic} = $isStatic;
    $self->{returnType} = $returnType;
    $self->{args} = $args;
    return $self;
}

sub call
{
    my ( $self, $obj, $args ) = @_;
    my $methodName = $self->{name};
    if ($self->{isStatic}) {
        my $method = $self->{cls}->{name}->can($methodName);
        return $method->(@{$args});
    } else {
        return $obj->$methodName(@{$args});
    }
}

package OneMethodArgument;

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    my ( $name, $type ) = @_;
    $self->{name} = $name;
    $self->{type} = $type;
    return $self;
}

package OneReflect;

our $classes = {};

sub get_class
{
    my ($obj) = @_;
    my $name = ref $obj;
    return get_class_by_name($name);
}

sub get_class_by_name
{
    my ($name) = @_;
    my $key = name_key($name);
    return $classes{$key};
}

sub setup_class
{
    my ($cls) = @_;
    my $key = name_key($cls->{name});
    $classes{$key} = $cls;
}

sub name_key
{
    my ($name) = @_;
    my $key = lc($name);
    $key =~ s/_//g;
    return $key;
}

1;